## LibRecommender models experiments

* https://github.com/massquantity/LibRecommender/tree/master/examples

* Try different models. 
* Note that we explicitly create negatives.
* Can do train test split by time or other things


* Data Format
    * One thing is important, the model assumes that user, item, and label column index are 0, 1, and 2, respectively. You may wish to change the column order if that's not the case.

Possible approahc + negatives - https://github.com/zhangruiskyline/DeepLearning/blob/master/doc/Recommendation.md#ranking 


* Negative sampling from the sparse user-item cooccurrence matrix
    * https://stackoverflow.com/questions/49971318/how-to-generate-negative-samples-in-tensorflow


In [1]:
# from catboost import CatBoostClassifier
from sklearn.preprocessing import LabelEncoder, StandardScaler, MinMaxScaler
import pandas as pd
from sklearn.model_selection import train_test_split, GroupShuffleSplit
import numpy as np
import random
np.random.seed(100)
random.seed(100)
from sklearn.linear_model import LogisticRegression, LogisticRegressionCV
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder, OrdinalEncoder
from sklearn.model_selection import train_test_split, GridSearchCV, GroupShuffleSplit

from sklearn.preprocessing import LabelEncoder
from sklearn.metrics import roc_auc_score, classification_report

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.metrics import TopKCategoricalAccuracy, Precision, SparseTopKCategoricalAccuracy # @4
from tensorflow.keras.utils import to_categorical
from tensorflow import feature_column
from tensorflow.keras import layers

%matplotlib inline

pd.set_option("display.max_columns", 90)

In [2]:
from libreco.data import split_by_ratio_chrono, DatasetFeat, split_by_ratio
from libreco.algorithms import (
    FM, WideDeep, DeepFM, AutoInt, DIN, YouTubeMatch, YouTubeRanking,ItemCF
)


Instructions for updating:
non-resource variables are not supported in the long term


In [3]:
def reset_state(name):
    tf.compat.v1.reset_default_graph()
    print("\n", "=" * 30, name, "=" * 30)

In [4]:
!nvidia-smi -L

GPU 0: GeForce RTX 2060 (UUID: GPU-d8cefda9-d4cb-990c-cc01-a2a4f2416484)


In [5]:
tf.compat.v1.keras.layers.enable_v2_dtype_behavior()
## https://www.tensorflow.org/guide/mixed_precision ## TF mixed precision - pytorch requires other setup
from tensorflow.keras.mixed_precision import experimental as mixed_precision

policy = mixed_precision.Policy('mixed_float16')
mixed_precision.set_policy(policy)
## will need to correct in places, e.g.: 
## outputs = layers.Activation('softmax', dtype='float32', name='predictions')(x)

The dtype policy mixed_float16 may run slowly because this machine does not have a GPU. Only Nvidia GPUs with compute capability of at least 7.0 run quickly with mixed_float16.
If you will use compatible GPU(s) not attached to this host, e.g. by running a multi-worker model, you can ignore this warning. This message will only be logged once
Instructions for updating:
Use tf.keras.mixed_precision.LossScaleOptimizer instead. LossScaleOptimizer now has all the functionality of DynamicLossScale


#### Features to add:
* Lag 
* Rank (popularity) of city, country (in general, +- given booker country)
* Count of hotel; user, trip size ? (may be leaky )
* Seasonal features - Holidays? , datetime

Aggregate feats:
* user changed country? last booking (lag 1) country change? 
* max/min/avg popularity rank of previous locations visited



We should create a dictionary of the rank, count, city/country etc' feats, so we can easily merge them when making more "negative" samples/feats for ranking.


* Consider using a df2 of df without dates + drop_duplicates, +- without user/trip id (After calcing that) .


Leaky or potentially leaky (Dependso n test set): 
* Target freq features - frequency of target city, given source county +- affiliate +- month of year +- given country (and interactions of target freq). 
    * Risk of leaks - depends of test data has temporal split or not. 
    * cartboost can do target encode, but this lets us do it for interactions, e.g. target city freq given the 2 countries and affiliate.
    * beware overfitting! 

In [6]:
MIN_TARGET_FREQ = 25 # drop target/city_id values that appear less than this many times, as final step's target 
KEEP_TOP_K_TARGETS = 0 # keep K most frequent city ID targets (redundnat with the above, )

## (some) categorical variables that appear less than this many times will be replaced with a placeholder value!
## Includes CITY id (but done after target filtering, to avoid creating a "rare class" target:
LOW_COUNT_THRESH = 10

RUN_TABNET = False
max_epochs = 15

GET_COUNT_AGG_FEATS = False ## disable getting count, rank etc' groupby features , for speedup

DROP_FIRST_ROW =  True #False  ## drop first interaction per user from train data

## for matrix factorization/CF:
### morte possible ID_cols :  # last (last step in trip) - would double data per user incorrectly
### hotel_country_lag1 , city_id_lag1  (very relevant - needs shared embeddingm and would increase cardinality a lot.. ) 
ID_COLS = [
#     'device_class',
#            'affiliate_id',
           'booker_country',
           'checkin_quarter',
#            "last",
          "first_hotel_country"] 
MF_KEEP_COLS = ["ID"]+ID_COLS+['city_id',"hotel_country"]

SAVE_TO_DISK = False

TARGET_COL =  'city_id' #"city_id"#'hotel_country' 
USER_ID_COL = "utrip_id"

In [7]:
# most basic categorical columns , without 'user_id', , 'utrip_id' ordevice_class - used for count encoding/filtering
BASE_CAT_COLS = ['city_id',  'affiliate_id', 'booker_country', 'hotel_country']

# ### features to get lags for. Not very robust. May want different feats for lags before -1
# LAG_FEAT_COLS = ['city_id', 'device_class',
#        'affiliate_id', 'booker_country', 'hotel_country', 
#        'duration', 'same_country', 'checkin_day', 'checkin_weekday',
#        'checkin_week',
#         'checkout_weekday','checkout_week',
#        'city_id_count', 'affiliate_id_count',
#        'booker_country_count', 'hotel_country_count', 
#        'checkin_month_count', 'checkin_week_count', 'city_id_nunique',
#        'affiliate_id_nunique', 'booker_country_nunique',
#        'hotel_country_nunique', 'city_id_rank_by_hotel_country',
#        'city_id_rank_by_booker_country', 'city_id_rank_by_affiliate',
#        'affiliate_id_rank_by_hotel_country',
#        'affiliate_id_rank_by_booker_country', 'affiliate_id_rank_by_affiliate',
#        'booker_country_rank_by_hotel_country',
#        'booker_country_rank_by_booker_country',
#        'booker_country_rank_by_affiliate',
#        'hotel_country_rank_by_hotel_country',
#        'hotel_country_rank_by_booker_country',
#        'hotel_country_rank_by_affiliate',
#        'checkin_month_rank_by_hotel_country',
#        'checkin_month_rank_by_booker_country',
#        'checkin_month_rank_by_affiliate']

In [8]:
# https://stackoverflow.com/questions/33907537/groupby-and-lag-all-columns-of-a-dataframe
# https://stackoverflow.com/questions/62924987/lag-multiple-variables-grouped-by-columns
## lag features with groupby over many columns: 
def groupbyLagFeatures(df:pd.DataFrame,lag:[]=[1,2],group="utrip_id",lag_feature_cols=[]):
    """
    lag features with groupby over many columns.
    Assumes sorted data!
    https://stackoverflow.com/questions/62924987/lag-multiple-variables-grouped-by-columns"""
    if len(lag_feature_cols)>0:
        df=pd.concat([df]+[df.groupby(group)[lag_feature_cols].shift(x).add_prefix('lag'+str(x)+"_") for x in lag],axis=1)
    else:
         df=pd.concat([df]+[df.groupby(group).shift(x).add_prefix('lag'+str(x)+"_") for x in lag],axis=1)
    return df


def groupbyFirstLagFeatures(df:pd.DataFrame,group="user_id",lag_feature_cols=[]):
    """
    Get  first/head value lag-like of features with groupby over columns. Assumes sorted data!
    """
    if len(lag_feature_cols)>0:
        df=pd.concat([df]+[df.groupby(group)[lag_feature_cols].transform("first").add_prefix("first_")],axis=1)
    else:
#          df=pd.concat([df]+[df.groupby(group).first().add_prefix("first_")],axis=1)
        df=pd.concat([df]+[df.groupby(group).transform("first").add_prefix("first_")],axis=1)
    return df

######## Get n most popular items, per group
def most_popular(group, n_max=4):
    """Find most popular hotel clusters by destination
    Define a function to get most popular hotels for a destination group.

    Previous version used nlargest() Series method to get indices of largest elements. But the method is rather slow.
    Source: https://www.kaggle.com/dvasyukova/predict-hotel-type-with-pandas
    """
    relevance = group['relevance'].values
    hotel_cluster = group['hotel_cluster'].values
    most_popular = hotel_cluster[np.argsort(relevance)[::-1]][:n_max]
    return np.array_str(most_popular)[1:-1] # remove square brackets


# A utility method to create a tf.data dataset from a Pandas Dataframe
def df_to_dataset(dataframe, shuffle=True, batch_size=64,target_col="target"):
    """
    Wrap dataframes with tf.data. 
    This will enable us to use feature columns as a bridge to map from the columns in a dataframe to features used to train the model.
    https://www.tensorflow.org/tutorials/structured_data/feature_columns#create_an_input_pipeline_using_tfdata
    """
    dataframe = dataframe.copy()
    labels = dataframe.pop(target_col)
    ds = tf.data.Dataset.from_tensor_slices((dict(dataframe), labels))
    if shuffle:
        ds = ds.shuffle(buffer_size=len(dataframe))
    ds = ds.batch(batch_size)
    return ds


import multiprocessing
import numpy as np
import pandas as pd


In [9]:
df = pd.read_csv("booking_train_set.csv",
#                  nrows=123456,
                 index_col=[0],
                 parse_dates=["checkin"],infer_datetime_format=True)

df.sort_values([ "checkin",
                "user_id"],inplace=True)
print(df.nunique())
df

user_id           200153
checkin              425
checkout             425
city_id            39901
device_class           3
affiliate_id        3254
booker_country         5
hotel_country        195
utrip_id          217686
dtype: int64


user_id    checkin    checkout  city_id device_class  affiliate_id  \
519017    727105 2015-12-31  2016-01-01    18820       mobile           359   
986596   2000964 2015-12-31  2016-01-01    63341       mobile          8151   
7504     2595109 2015-12-31  2016-01-01    27404       mobile           359   
52884     110418 2016-01-01  2016-01-02     3763      desktop          9924   
1068508   221863 2016-01-01  2016-01-02    45597       mobile          7774   
...          ...        ...         ...      ...          ...           ...   
908844   5755992 2017-02-27  2017-02-28    52860      desktop          4568   
1154024  5842454 2017-02-27  2017-02-28    35850      desktop           384   
1160065  5936647 2017-02-27  2017-02-28    20199      desktop          4541   
244942   5955565 2017-02-27  2017-02-28    54384      desktop          8132   
787249   6172320 2017-02-27  2017-02-28     6005       mobile          9452   

               booker_country hotel_country   utrip_id  
519017   The Devilfire Empire  Cobra Island   727105_1  
986596   The Devilfire Empire  Cobra Island  2000964_1  
7504     The Devilfire Empire  Cobra Island  2595109_1  
52884    The Devilfire Empire  Glubbdubdrib   110418_1  
1068508                Gondal        Gondal   221863_1  
...                       ...           ...        ...  
908844                 Gondal       Axphain  5755992_4  
1154024                Gondal  Rook Islands  5842454_6  
1160065                Gondal        Kasnia  5936647_2  
244942                Elbonia       Patusan  5955565_2  
787249   The Devilfire Empire       Patusan  6172320_1  

[1166835 rows x 9 columns]

In [10]:
### i disabled most of thefeature eztraction here for simplicity

# df["duration"] = (df["checkout"] - df["checkin"]).dt.days
# df["same_country"] = (df["booker_country"]==df["hotel_country"]).astype(int)

# df["checkin_day"] = df["checkin"].dt.day
# df["checkin_weekday"] = df["checkin"].dt.weekday
df["checkin_week"] = df["checkin"].dt.isocalendar().week.astype(int) ## week of year
df["checkin_month"] = df["checkin"].dt.month
# df["checkin_year"] = df["checkin"].dt.year-2016

df["checkin_quarter"] = df["checkin"].dt.quarter # relatively redundant but may be used for "id"

# df["checkin_quarter"] = df["checkin_quarter"]/4 # scale. could also do cos, sin extraction. makesi t a float instead of int/embedding


# df["checkout_weekday"] = df["checkout"].dt.weekday
# df["checkout_week"] = df["checkout"].dt.isocalendar().week.astype(int) ## week of year
# df["checkout_day"] = df["checkout"].dt.day ## day of month

## cyclical datetime embeddings
## drop originakl variables? 
## TODO:L add for other variables, +- those that we'll embed (week?)

# df['checkin_weekday_sin'] = np.sin(df["checkin_weekday"]*(2.*np.pi/7))
# df['checkin_weekday_cos'] = np.cos(df["checkin_weekday"]*(2.*np.pi/7))
# df['checkin_month_sin'] = np.sin((df["checkin_month"]-1)*(2.*np.pi/12))
# df['checkin_month_cos'] = np.cos((df["checkin_month"]-1)*(2.*np.pi/12))
df['checkin_week_sin'] = np.sin((df["checkin_week"]-1)*(2.*np.pi/53))
df['checkin_week_cos'] = np.cos((df["checkin_week"]-1)*(2.*np.pi/53))


# #############
# # last number in utrip id - probably which trip number it is:
# df["utrip_number"] = df["utrip_id"].str.split("_",expand=True)[1].astype(int)

### encode string columns - must be consistent with test data 
### IF we can concat test with train, we can just do a single transformation  for the NON TARGET cols
# obj_cols_list = df.select_dtypes("O").columns.values
obj_cols_list = ['device_class','booker_country','hotel_country',
#                 "city_id"
                ] # we could also define when loading data, dtype

# for c in obj_cols_list:
#     df[c] = df[c].astype("category")
#     df[c] = df[c].cat.codes.astype(int)
#     print("min",df[c].min()) # min is 0 - which is what the embedding (indices) will expect

## view steps of a trip per user & trip, in order. ## last step == 1.
## count #/pct step in a trip (utrip_id) per user. Useful to get the "final" step per trip - for prediction
## note that the order is ascending, so we would need to select by "last" . (i.e "1" is the first step, 2 the second, etc') , or we could use pct .rank(ascending=True,pct=True)
#### this feature overlaps with the count of each trip id (for the final row)
##  = df.sort_values(["checkin","checkout"])... - df already sorted above
# df["utrip_steps_from_end"] = df.groupby("utrip_id")["checkin"].rank(ascending=True,pct=True) 

In [11]:
### add features to be consistent with test set of row in trip, and total trips in trip

df["row_num"] = df.groupby("utrip_id")["checkin"].rank(ascending=True,pct=False).astype(int)
utrip_counts = df["utrip_id"].value_counts()
df["total_rows"] = df["utrip_id"].map(utrip_counts)

### last step in trip
df["last"] = (df["total_rows"]==df["row_num"]).astype(int)

df["total_rows"].describe();

### replace rare categorical variable(s) - affiliates
* replace rare variables (under 2 occurrences) with "-1" dummy

In [12]:
### replace rare variables (under 2 occurrences) with "-1" dummy

affiliates_counts = df["affiliate_id"].value_counts()
print("before:", affiliates_counts)
print("uniques",df["affiliate_id"].nunique())
affiliates_counts = affiliates_counts.to_dict()
# df["affiliate_id"] = df["affiliate_id"].where(df["affiliate_id"].apply(lambda x: x.map(x.value_counts()))>=3, -1)
df["affiliate_id"] = df["affiliate_id"].where(df["affiliate_id"].map(affiliates_counts)>=3, -2)
df["affiliate_id"] = df["affiliate_id"].astype(int)

print("after\n",df["affiliate_id"].value_counts())
print("uniques",df["affiliate_id"].nunique())

before: 9924     277775
359      171385
384       88137
9452      85476
4541      41504
          ...  
8351          1
8464          1
2202          1
10513         1
2047          1
Name: affiliate_id, Length: 3254, dtype: int64
uniques 3254
after
 9924    277775
359     171385
384      88137
9452     85476
4541     41504
         ...  
2615         3
5963         3
2618         3
838          3
176          3
Name: affiliate_id, Length: 2152, dtype: int64
uniques 2152


In [13]:
# df["total_rows"].map(pd.cut(df["total_rows"],bins=3,))

#### Add LAG feature(s)  + DROP First visited location rows
* "first" hotel country (vs most recent country visited)
    * `groupbyFirstLagFeatures(df,group="user_id",lag_feature_cols=["hotel_country","city_id"])`
* Can consider: lag1 hotel_country, hotel_id 

In [14]:
## add the "first" place visited/values
### nopte - will need to drop first row in trip, or impute nans when using this feature 

### first by user results in too much sparsity/rareness for our IDs purposes
# df = groupbyFirstLagFeatures(df,group="user_id",lag_feature_cols=["hotel_country"]) # ["hotel_country","city_id"]

## alt - messy, but maybe good enough : 
# df = groupbyFirstLagFeatures(df,group=['device_class', 'affiliate_id',
#                                        'booker_country','checkin_month',"last"],
#                              lag_feature_cols=["hotel_country","city_id",'device_class', 'affiliate_id',
#                                               "checkin_quarter","checkin_month","booker_country"])

df = groupbyFirstLagFeatures(df,group=['utrip_id'],
                             lag_feature_cols=["hotel_country","city_id",'device_class', 'affiliate_id',
                                              "checkin_month","booker_country"])

print(df[["first_hotel_country","hotel_country","city_id","first_city_id"]].nunique())

first_hotel_country      174
hotel_country            195
city_id                39901
first_city_id          19674
dtype: int64


In [15]:
## add lag features
df = groupbyLagFeatures(df,lag=[1,2],lag_feature_cols=["hotel_country","city_id"])
## HAcky: reparse as integers
df = df.fillna(-1) ## might be best to fill with max + 1 or something ? negative may cause error ? 
df = df.infer_objects()
df["lag1_city_id"] = df["lag1_city_id"].astype(int)
df["lag2_city_id"] = df["lag2_city_id"].astype(int)
df

user_id    checkin    checkout  city_id device_class  affiliate_id  \
519017    727105 2015-12-31  2016-01-01    18820       mobile           359   
986596   2000964 2015-12-31  2016-01-01    63341       mobile          8151   
7504     2595109 2015-12-31  2016-01-01    27404       mobile           359   
52884     110418 2016-01-01  2016-01-02     3763      desktop          9924   
1068508   221863 2016-01-01  2016-01-02    45597       mobile          7774   
...          ...        ...         ...      ...          ...           ...   
908844   5755992 2017-02-27  2017-02-28    52860      desktop          4568   
1154024  5842454 2017-02-27  2017-02-28    35850      desktop           384   
1160065  5936647 2017-02-27  2017-02-28    20199      desktop          4541   
244942   5955565 2017-02-27  2017-02-28    54384      desktop          8132   
787249   6172320 2017-02-27  2017-02-28     6005       mobile          9452   

               booker_country hotel_country   utrip_id  checkin_week  \
519017   The Devilfire Empire  Cobra Island   727105_1            53   
986596   The Devilfire Empire  Cobra Island  2000964_1            53   
7504     The Devilfire Empire  Cobra Island  2595109_1            53   
52884    The Devilfire Empire  Glubbdubdrib   110418_1            53   
1068508                Gondal        Gondal   221863_1            53   
...                       ...           ...        ...           ...   
908844                 Gondal       Axphain  5755992_4             9   
1154024                Gondal  Rook Islands  5842454_6             9   
1160065                Gondal        Kasnia  5936647_2             9   
244942                Elbonia       Patusan  5955565_2             9   
787249   The Devilfire Empire       Patusan  6172320_1             9   

         checkin_month  checkin_quarter  checkin_week_sin  checkin_week_cos  \
519017              12                4         -0.118273          0.992981   
986596              12                4         -0.118273          0.992981   
7504                12                4         -0.118273          0.992981   
52884                1                1         -0.118273          0.992981   
1068508              1                1         -0.118273          0.992981   
...                ...              ...               ...               ...   
908844               2                1          0.812487          0.582979   
1154024              2                1          0.812487          0.582979   
1160065              2                1          0.812487          0.582979   
244942               2                1          0.812487          0.582979   
787249               2                1          0.812487          0.582979   

         row_num  total_rows  last first_hotel_country  first_city_id  \
519017         1           4     0        Cobra Island          18820   
986596         1           5     0        Cobra Island          63341   
7504           1           4     0        Cobra Island          27404   
52884          1          10     0        Glubbdubdrib           3763   
1068508        1           8     0              Gondal          45597   
...          ...         ...   ...                 ...            ...   
908844         5           5     1              Gondal          36063   
1154024        9           9     1         Kumbolaland          18032   
1160065       12          12     1              Kasnia           2078   
244942         5           5     1             Patusan          60966   
787249        10          10     1             Patusan          60143   

        first_device_class  first_affiliate_id  first_checkin_month  \
519017              mobile                 359                   12   
986596              mobile                8151                   12   
7504                mobile                 359                   12   
52884              desktop                9924                    1   
1068508       

In [16]:
df.dtypes

user_id                          int64
checkin                 datetime64[ns]
checkout                        object
city_id                          int64
device_class                    object
affiliate_id                     int32
booker_country                  object
hotel_country                   object
utrip_id                        object
checkin_week                     int32
checkin_month                    int64
checkin_quarter                  int64
checkin_week_sin               float64
checkin_week_cos               float64
row_num                          int32
total_rows                       int64
last                             int32
first_hotel_country             object
first_city_id                    int64
first_device_class              object
first_affiliate_id               int32
first_checkin_month              int64
first_booker_country            object
lag1_hotel_country              object
lag1_city_id                     int32
lag2_hotel_country       

In [17]:
df.nunique()

user_id                 200153
checkin                    425
checkout                   425
city_id                  39901
device_class                 3
affiliate_id              2152
booker_country               5
hotel_country              195
utrip_id                217686
checkin_week                53
checkin_month               12
checkin_quarter              4
checkin_week_sin            53
checkin_week_cos            45
row_num                     48
total_rows                  41
last                         2
first_hotel_country        174
first_city_id            19674
first_device_class           3
first_affiliate_id        1916
first_checkin_month         12
first_booker_country         5
lag1_hotel_country         192
lag1_city_id             37469
lag2_hotel_country         184
lag2_city_id             34097
dtype: int64

* Continue

In [18]:
##### Following aggregation features - would be best to use time window (sort data) to generate, otherwise they will LEAK! (e.g. nunique countries visited)

### count features (can also later add rank inside groups).
### Some may be leaks (# visits in a trip should use time window?) , and do users repeat? 
### can add more counts of group X time period (e.g. affiliate X month of year)

## alt way to get counts/freq :

if GET_COUNT_AGG_FEATS:
    count_cols = [ 'city_id','affiliate_id', 'booker_country', 'hotel_country', 
    #               'utrip_id','user_id', 
     "checkin_month","checkin_week"]
    for c in count_cols:
        df[f"{c}_count"] = df.groupby([c])["duration"].transform("size")

    ########################################################
    ## nunique per trip
    ### https://stackoverflow.com/questions/46470743/how-to-efficiently-compute-a-rolling-unique-count-in-a-pandas-time-series

    nunique_cols = [ 'city_id','affiliate_id', 'booker_country', 'hotel_country']
    # df["nunique_booker_countries"] = df.groupby("utrip_id")["booker_country"].nunique()
    # df["nunique_hotel_country"] = df.groupby("utrip_id")["hotel_country"].nunique()
    for c in nunique_cols:
        df[f"{c}_nunique"] = df.groupby(["utrip_id"])[c].transform("nunique")
    print(df.nunique())

    ########################################################
    ## get frequency/count feature's rank within a group - e.g. within a country (or affiliate) 
    ## add "_count" to column name to get count col name, then add rank col 

    ### ALT/ duplicate feat - add percent rank (instead or in addition)

    rank_cols = ['city_id','affiliate_id', 'booker_country','hotel_country',
     "checkin_month"]
    ### what is meaning of groupby and rank of smae variable by same var? Surely should be 1 / unary? 
    for c in rank_cols:
        df[f"{c}_rank_by_hotel_country"] = df.groupby(['hotel_country'])[f"{c}_count"].transform("rank")
        df[f"{c}_rank_by_booker_country"] = df.groupby(['booker_country'])[f"{c}_count"].transform("rank")
        df[f"{c}_rank_by_affiliate"] = df.groupby(['affiliate_id'])[f"{c}_count"].transform("rank")     
else:
    freq = df["city_id"].value_counts()
    df["city_id_count"] = df["city_id"].map(freq)
    print(freq.describe())
    
df

count    39901.000000
mean        29.243252
std        218.801654
min          1.000000
25%          1.000000
50%          3.000000
75%          9.000000
max      11242.000000
Name: city_id, dtype: float64


user_id    checkin    checkout  city_id device_class  affiliate_id  \
519017    727105 2015-12-31  2016-01-01    18820       mobile           359   
986596   2000964 2015-12-31  2016-01-01    63341       mobile          8151   
7504     2595109 2015-12-31  2016-01-01    27404       mobile           359   
52884     110418 2016-01-01  2016-01-02     3763      desktop          9924   
1068508   221863 2016-01-01  2016-01-02    45597       mobile          7774   
...          ...        ...         ...      ...          ...           ...   
908844   5755992 2017-02-27  2017-02-28    52860      desktop          4568   
1154024  5842454 2017-02-27  2017-02-28    35850      desktop           384   
1160065  5936647 2017-02-27  2017-02-28    20199      desktop          4541   
244942   5955565 2017-02-27  2017-02-28    54384      desktop          8132   
787249   6172320 2017-02-27  2017-02-28     6005       mobile          9452   

               booker_country hotel_country   utrip_id  checkin_week  \
519017   The Devilfire Empire  Cobra Island   727105_1            53   
986596   The Devilfire Empire  Cobra Island  2000964_1            53   
7504     The Devilfire Empire  Cobra Island  2595109_1            53   
52884    The Devilfire Empire  Glubbdubdrib   110418_1            53   
1068508                Gondal        Gondal   221863_1            53   
...                       ...           ...        ...           ...   
908844                 Gondal       Axphain  5755992_4             9   
1154024                Gondal  Rook Islands  5842454_6             9   
1160065                Gondal        Kasnia  5936647_2             9   
244942                Elbonia       Patusan  5955565_2             9   
787249   The Devilfire Empire       Patusan  6172320_1             9   

         checkin_month  checkin_quarter  checkin_week_sin  checkin_week_cos  \
519017              12                4         -0.118273          0.992981   
986596              12                4         -0.118273          0.992981   
7504                12                4         -0.118273          0.992981   
52884                1                1         -0.118273          0.992981   
1068508              1                1         -0.118273          0.992981   
...                ...              ...               ...               ...   
908844               2                1          0.812487          0.582979   
1154024              2                1          0.812487          0.582979   
1160065              2                1          0.812487          0.582979   
244942               2                1          0.812487          0.582979   
787249               2                1          0.812487          0.582979   

         row_num  total_rows  last first_hotel_country  first_city_id  \
519017         1           4     0        Cobra Island          18820   
986596         1           5     0        Cobra Island          63341   
7504           1           4     0        Cobra Island          27404   
52884          1          10     0        Glubbdubdrib           3763   
1068508        1           8     0              Gondal          45597   
...          ...         ...   ...                 ...            ...   
908844         5           5     1              Gondal          36063   
1154024        9           9     1         Kumbolaland          18032   
1160065       12          12     1              Kasnia           2078   
244942         5           5     1             Patusan          60966   
787249        10          10     1             Patusan          60143   

        first_device_class  first_affiliate_id  first_checkin_month  \
519017              mobile                 359                   12   
986596              mobile                8151                   12   
7504                mobile                 359                   12   
52884              desktop                9924                    1   
1068508       

In [19]:
print("cities with >= 7 occurences:",df.loc[df["city_id_count"]>=7]["city_id"].nunique())
print(f"cities with >= MIN_TARGET_FREQ ({MIN_TARGET_FREQ}) occurences:",df.loc[df["city_id_count"]>=MIN_TARGET_FREQ]["city_id"].nunique())
print(f"rows left if filtering by MIN_TARGET_FREQ :",df.loc[df["city_id_count"]>=MIN_TARGET_FREQ].shape[0])

cities with >= 7 occurences: 12345
cities with >= MIN_TARGET_FREQ (25) occurences: 4822
rows left if filtering by MIN_TARGET_FREQ : 1010884


In [20]:
assert df.isna().sum().max() ==0

In [21]:
# df2 = df.drop_duplicates(subset=["user_id","city_id"])["city_id"].copy()
# print(df2.shape[0])
# print("df2 nunique (cities without duplicate user visits)",df2.nunique())

# c2_counts = df2.value_counts()

# print("city counts")
# print(c2_counts)
# print(c2_counts.describe())
# print("cities with at least 2:",(c2_counts>=2).sum())
# print("cities with at least 3:",(c2_counts>=3).sum())
# print("cities with at least 5:",(c2_counts>=5).sum())
# print("cities with at least 10:",(c2_counts>=10).sum())
# print("cities with at least 15:",(c2_counts>=15).sum())
# print("cities with at least 20:",(c2_counts>=20).sum())
# print("cities with at least 30:",(c2_counts>=30).sum())
# print("cities with at least 50:",(c2_counts>=50).sum())
# print("cities with at least 100:",(c2_counts>=100).sum())

# c2_freq = df2.value_counts(normalize=True)
# print("top 4 sum coverage (normalized): ",c2_freq[0:4].sum().round(3))
# print("top 50 sum coverage (normalized): ",c2_freq[0:50].sum().round(3))
# print("top 100 sum coverage (normalized): ",c2_freq[0:100].sum().round(3))
# print("top 400 sum coverage (normalized): ",c2_freq[0:400].sum().round(3))
# print("top 1,000 sum coverage (normalized): ",c2_freq[0:1000].sum().round(3))
# print("top 5,000 sum coverage (normalized): ",c2_freq[0:5000].sum().round(3))
# print("top 8,000 sum coverage (normalized): ",c2_freq[0:8000].sum().round(3))
# print("top 15,000 sum coverage (normalized): ",c2_freq[0:15000].sum().round(3))

## Frequent city target List + City count encoding
* Get the K most frequent target city IDs - selected based on frequency as final destination (not just overall)
* +- Also after this, replace rare city IDs categorical features with count encoding to reduce dimensionality
    * Keep them as count, or aggregate all of them as "under_K"?

##### Output  : `TOP_TARGETS` - filter data by this *after* creation of lag features ! 

* Drop duplicates by the same user (reduce possible bias of frequent users? Only relevant if test is seperater from "frequent travellers") 
    * results in 216,633 , vs 217,686 without dropping duplicates by users
    * ~19.9k unique cities
    
* Could do other encodings - https://contrib.scikit-learn.org/category_encoders/count.html

* Note that all this is after we've added rank, count features beforehand, so that information won't be lost for these variables, despite these transforms



* **NOTE** he most frequent final destinations are NOT the same as the most popular overall destinations +- first location ! 

In [22]:
if KEEP_TOP_K_TARGETS > 0 :
    df_end = df.loc[df["utrip_steps_from_end"]==1].drop_duplicates(subset=["city_id","hotel_country","user_id"])[["city_id","hotel_country"]].copy()
    print(df_end.shape[0])
    end_city_counts = df_end.city_id.value_counts()
    print(end_city_counts)
    
    TOP_TARGETS = end_city_counts.head(KEEP_TOP_K_TARGETS).index.values
    print(f"top {KEEP_TOP_K_TARGETS} targets \n",TOP_TARGETS)
    
#     assert df.loc[df["city_id"].isin(TOP_TARGETS)]["city_id"].nunique() == KEEP_TOP_K_TARGETS

####        
# replace low frequency categoircal features    

# ##replace with count encoding if have at least k, group rarest as "-1":# df[BASE_CAT_COLS] = df[BASE_CAT_COLS].where(df[BASE_CAT_COLS].apply(lambda x: x.map(x.value_counts()))>=LOW_COUNT_THRESH, -1)   
# ## replace/group only the rare variables : 
# df[BASE_CAT_COLS] = df[BASE_CAT_COLS].where(df[BASE_CAT_COLS].apply(lambda x: x.map(x.value_counts()))>=LOW_COUNT_THRESH, -1)
# df[BASE_CAT_COLS].head()

### Filter data by city id frequency
#### df2 - smaller df (may not ben ecessary to make

* drop rows if it's city id appears less than X times - this is prior to CF
* We could also add inclusion/exclusion based on target appearing/frequency as target in final stage of rtip - optional
* maybe also drop (end exclude in freq counting) thefirst point in a trip ? 



In [23]:
### unsure about this filtering - depends if data points are real or mistake
print("dropping users with less than 4 trips")
df = df.loc[df["total_rows"]>=4]#.copy()
# print("abnormal users dropped",df.shape[0]-df.shape[0])

print(f"dropping cities with less than {MIN_TARGET_FREQ} occurences:")
df = df.loc[df.groupby(["city_id"])["hotel_country"].transform("count")>=MIN_TARGET_FREQ] ## update count
# df = df.loc[df["city_id_count"]>=MIN_TARGET_FREQ]
print(df.shape[0])

# print(f"dropping users with less than 4 instances, after previous city filter:")
df = df.loc[df.groupby(["utrip_id"])[TARGET_COL].transform("count")>=4]

# print(f"dropping cities with less than {MIN_TARGET_FREQ} occurences:")
df = df.loc[df.groupby(["city_id"])["hotel_country"].transform("count")>=MIN_TARGET_FREQ]

print(f"dropping users with less than 4 instances, after previous filters:")
df = df.loc[df.groupby(["utrip_id"])[TARGET_COL].transform("count")>=4]

df = df.loc[df.groupby(["city_id"])["hotel_country"].transform("count")>=MIN_TARGET_FREQ]
df = df.loc[df.groupby(["utrip_id"])[TARGET_COL].transform("count")>=4]
df = df.loc[df.groupby(["city_id"])["hotel_country"].transform("count")>=MIN_TARGET_FREQ]
df = df.loc[df.groupby(["utrip_id"])[TARGET_COL].transform("count")>=4]
# df = df.loc[df.groupby(["city_id"])["hotel_country"].transform("count")>=MIN_TARGET_FREQ]
# df = df.loc[df.groupby(["utrip_id"])[TARGET_COL].transform("count")>=4]

print("rows left:",df.shape[0])

print("nunique cities after freq filt",df[["city_id","utrip_id","user_id","hotel_country"]].nunique())
print("nunique city_id per hotel_country:")
df.groupby(["hotel_country"])["city_id"].nunique().describe()

dropping users with less than 4 trips
dropping cities with less than 25 occurences:
1008673
dropping users with less than 4 instances, after previous filters:
rows left: 857829
nunique cities after freq filt city_id            3667
utrip_id         162904
user_id          152608
hotel_country       105
dtype: int64
nunique city_id per hotel_country:


count    105.00000
mean      34.92381
std       69.87058
min        1.00000
25%        2.00000
50%       10.00000
75%       34.00000
max      429.00000
Name: city_id, dtype: float64

* drop the first location visited per trip, as that shares the "first country/hotel/city id feature" ??

In [24]:
if DROP_FIRST_ROW: ## drop the first location visited per trip, (row_num = 1) from df data/interactions
    df = df.loc[df["row_num"]>1]
    print(df.shape[0])

702158


In [25]:
df.sort_values([ "checkin",
                "utrip_id"],inplace=True)

In [26]:
DROP_COLS = [ 'user_id', 'checkout',
            'row_num',"last"] ## dr

In [27]:
df["label"] = 1

## rearrange columns - user, item, and label column index are 0, 1, and 2
# new_columns = cols_to_order + (frame.columns.drop(cols_to_order).tolist())
cols_to_order = ["utrip_id","city_id","label","checkin"]
new_columns = cols_to_order + (df.columns.drop(cols_to_order+DROP_COLS).tolist())

df = df[new_columns]
print(df.columns)
# df.tail(2)

Index(['utrip_id', 'city_id', 'label', 'checkin', 'device_class',
       'affiliate_id', 'booker_country', 'hotel_country', 'checkin_week',
       'checkin_month', 'checkin_quarter', 'checkin_week_sin',
       'checkin_week_cos', 'total_rows', 'first_hotel_country',
       'first_city_id', 'first_device_class', 'first_affiliate_id',
       'first_checkin_month', 'first_booker_country', 'lag1_hotel_country',
       'lag1_city_id', 'lag2_hotel_country', 'lag2_city_id', 'city_id_count'],
      dtype='object')


In [28]:
KEEP_COLS = ['utrip_id', 'city_id', 'label',
             "checkin",
             "hotel_country", ## should be item feature 
             
#               'user_id', 'checkin', 'checkout','device_class', 'affiliate_id',  'booker_country', 
#        'checkin_week', 'checkin_month', 
              'row_num', 
             'total_rows',
#               'last',
       'first_hotel_country',
             'first_city_id', 'first_device_class','first_affiliate_id',
#              'first_checkin_quarter',
#              'first_checkin_month',
       'first_booker_country'
            ]

df_feat = df.copy()   
# df_feat = df_feat[KEEP_COLS] 

df_feat.rename(columns={"checkin":"time",
                  'utrip_id':"user",
                   'city_id':"item"
                  },inplace=True)
print(df_feat.shape)
# ### opt: drop duplicates, +- time ? 
# df_feat.drop_duplicates(subset=['user', 'item', 'hotel_country', 'first_hotel_country',
#                            'first_city_id', 'first_booker_country'],inplace=True)

print(df_feat.shape[0])
print(df_feat.columns.drop(cols_to_order,errors="ignore").tolist())
df_feat

'device_class', 'affiliate_id', 'booker_country',

(702158, 25)
702158
['user', 'item', 'time', 'device_class', 'affiliate_id', 'booker_country', 'hotel_country', 'checkin_week', 'checkin_month', 'checkin_quarter', 'checkin_week_sin', 'checkin_week_cos', 'total_rows', 'first_hotel_country', 'first_city_id', 'first_device_class', 'first_affiliate_id', 'first_checkin_month', 'first_booker_country', 'lag1_hotel_country', 'lag1_city_id', 'lag2_hotel_country', 'lag2_city_id', 'city_id_count']


('device_class', 'affiliate_id', 'booker_country')

In [29]:
print("object cols:")
print(list(df_feat.select_dtypes("O").columns))
print()
print("numerical cols:")
print(list(df_feat.select_dtypes(exclude="O").columns))


object cols:
['user', 'device_class', 'booker_country', 'hotel_country', 'first_hotel_country', 'first_device_class', 'first_booker_country', 'lag1_hotel_country', 'lag2_hotel_country']

numerical cols:
['item', 'label', 'time', 'affiliate_id', 'checkin_week', 'checkin_month', 'checkin_quarter', 'checkin_week_sin', 'checkin_week_cos', 'total_rows', 'first_city_id', 'first_affiliate_id', 'first_checkin_month', 'lag1_city_id', 'lag2_city_id', 'city_id_count']


In [30]:
sparse_columns = ['device_class', 'booker_country', 'hotel_country',
       'first_hotel_country', 'first_device_class', 'first_booker_country',
        'affiliate_id','first_city_id', 'first_affiliate_id',
              'lag1_city_id','lag2_city_id',
       'lag1_hotel_country', 'lag2_hotel_country'
             , 'checkin_week']

dense_columns = ['checkin_month', 'checkin_quarter', 'checkin_week_sin', 'checkin_week_cos',
#              'row_num', 
                 'total_rows',
#              'last',
             'first_city_id', 'first_affiliate_id', 'first_checkin_month', 'city_id_count']

In [31]:
df_feat[sparse_columns] = df_feat[sparse_columns].astype(str)
df_feat[sparse_columns].dtypes

device_class            object
booker_country          object
hotel_country           object
first_hotel_country     object
first_device_class      object
first_booker_country    object
affiliate_id            object
first_city_id           object
first_affiliate_id      object
lag1_city_id            object
lag2_city_id            object
lag1_hotel_country      object
lag2_hotel_country      object
checkin_week            object
dtype: object

In [32]:
df_feat[dense_columns].dtypes

checkin_month            int64
checkin_quarter          int64
checkin_week_sin       float64
checkin_week_cos       float64
total_rows               int64
first_city_id           object
first_affiliate_id      object
first_checkin_month      int64
city_id_count            int64
dtype: object

### model

In [33]:
print("Sparsity: (total positives/interactions, divided by all possible user/item combs in %, )")
# print(100*df.shape[0] / (162904 * 3667))
print(100*df_feat.shape[0] / (df_feat['user'].nunique() * df_feat['item'].nunique()))

Sparsity: (total positives/interactions, divided by all possible user/item combs in %, )
0.1175417602207954


In [34]:
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
os.environ["KMP_WARNINGS"] = "FALSE"
tf.compat.v1.logging.set_verbosity(tf.compat.v1.logging.ERROR)

In [35]:
reset_state("YouTubeRanking")
# reset_state("YouTubeMatch") 


 ============================== YouTubeRanking ==============================


In [36]:
%%time

# train_data, test_data = split_by_ratio_chrono(df, test_size=0.15)
train_data, test_data = split_by_ratio(df_feat, test_size=0.2)

# specify complete columns information
# sparse_col = ['first_hotel_country', 
# #               'first_city_id','first_device_class','first_affiliate_id', 
#               'first_booker_country',
#               'first_city_id', 'first_device_class','first_affiliate_id',
#              "hotel_country"]
# # dense_col = [] #['total_rows',
# #              'first_checkin_quarter', 
#               #'first_checkin_month',]
# user_col = ['first_hotel_country','first_booker_country','first_city_id', 'first_device_class','first_affiliate_id']

sparse_col = sparse_columns
dense_col = dense_columns
item_col = ["hotel_country"]
user_col = [i for i in sparse_columns+ dense_columns if i not in item_col] ## evrything except for hotel_country

# train_data, test_data = split_by_ratio(df[user_col+item_col+["label"]+sparse_col+dense_col], test_size=0.15)

# train_data, data_info = DatasetFeat.build_trainset(
#     train_data, user_col, item_col, sparse_col, dense_col
# )
train_data, data_info = DatasetFeat.build_trainset(
    train_data, item_col=item_col, sparse_col=sparse_col,user_col=user_col,
)
print("trainset made")
test_data = DatasetFeat.build_testset(test_data)
print("testset made")

# sample negative items for each record
train_data.build_negative_samples(data_info,num_neg=3)
print("train negs done")
test_data.build_negative_samples(data_info,num_neg=1)
print("test negs done")
print(data_info)  # n_users: 9058, n_items: 456, data sparsity: 0.7836 %
## (min 25 cities): n_users: 162904, n_items: 3667, data sparsity: 0.0977 %

# reset_state("YouTubeRanking")
# reset_state("YouTubeMatch") 
### The YouTubeMatch model assumes no item features.  

## YouTubeRanking  ## YouTubeMatch # DeepFM
# ytb_ranking = YouTubeRanking(task="ranking", data_info=data_info,
#                              embed_size=16, n_epochs=3, lr=1e-3,
#                              recent_num=10,
#                              batch_size=128, use_bn=True,
#                              hidden_units="128,64") ## hidden_units="128,64,32"

# ytb_ranking = DIN(task="ranking", data_info=data_info,
#                              n_epochs=3, lr=2e-3,
#                              recent_num=8,
# #                   use_tf_attention=True,
# #                   num_neg=2,
#                              batch_size=129, use_bn=True,
#                              hidden_units="128,64") ## hidden_units="128,64,32"
 
# ytb_ranking = AutoInt(task="ranking", data_info=data_info,
#                              n_epochs=15, lr=1e-3, batch_size=512,num_neg=3,num_heads=4,)

ytb_ranking = DeepFM(task="ranking", data_info=data_info,num_neg=3,)

ytb_ranking.fit(train_data, verbose=1, shuffle=True,
                eval_data=test_data,
                metrics=[ "loss", "roc_auc", "precision", "recall"]) # "loss", "roc_auc", "precision", "recall"


print(ytb_ranking.evaluate(test_data,metrics=[ "roc_auc", "precision", "recall"],k=4))
print("recall @90",ytb_ranking.evaluate(test_data,metrics=[ "recall"],k=90))

# # predict preference of user 1 to item 2333
# print("prediction: ", ytb_ranking.predict(user=1, item=33))
# # recommend 7 items for user 1
# print("recommendation: ", ytb_ranking.recommend_user(user=1, n_rec=3))

trainset made
testset made
random neg item sampling elapsed: 1.012s
train negs done
random neg item sampling elapsed: 0.092s
test negs done
n_users: 162904, n_items: 3667, data sparsity: 0.1005 %
Training start time: 2021-01-06 20:58:21


train: 100%|███████████████████████████████████████████████████████████████████████| 4691/4691 [03:43<00:00, 21.02it/s]


Epoch 1 elapsed: 224.468s


train: 100%|███████████████████████████████████████████████████████████████████████| 4691/4691 [03:54<00:00, 20.02it/s]


Epoch 2 elapsed: 235.525s


train: 100%|███████████████████████████████████████████████████████████████████████| 4691/4691 [03:44<00:00, 20.87it/s]


Epoch 3 elapsed: 225.824s


train: 100%|███████████████████████████████████████████████████████████████████████| 4691/4691 [03:56<00:00, 19.80it/s]


Epoch 4 elapsed: 238.138s


train: 100%|███████████████████████████████████████████████████████████████████████| 4691/4691 [04:02<00:00, 19.37it/s]


Epoch 5 elapsed: 243.490s


train: 100%|███████████████████████████████████████████████████████████████████████| 4691/4691 [03:57<00:00, 19.78it/s]


Epoch 6 elapsed: 238.399s


train: 100%|███████████████████████████████████████████████████████████████████████| 4691/4691 [03:50<00:00, 20.36it/s]


Epoch 7 elapsed: 232.185s


train: 100%|███████████████████████████████████████████████████████████████████████| 4691/4691 [03:49<00:00, 20.48it/s]


Epoch 8 elapsed: 230.082s


train: 100%|███████████████████████████████████████████████████████████████████████| 4691/4691 [03:41<00:00, 21.15it/s]


Epoch 9 elapsed: 222.859s


train: 100%|███████████████████████████████████████████████████████████████████████| 4691/4691 [03:49<00:00, 20.47it/s]


Epoch 10 elapsed: 231.046s


train: 100%|███████████████████████████████████████████████████████████████████████| 4691/4691 [04:01<00:00, 19.45it/s]


Epoch 11 elapsed: 242.184s


train: 100%|███████████████████████████████████████████████████████████████████████| 4691/4691 [03:41<00:00, 21.19it/s]


Epoch 12 elapsed: 222.535s


train: 100%|███████████████████████████████████████████████████████████████████████| 4691/4691 [03:54<00:00, 19.97it/s]


Epoch 13 elapsed: 235.946s


train: 100%|███████████████████████████████████████████████████████████████████████| 4691/4691 [04:02<00:00, 19.37it/s]


Epoch 14 elapsed: 243.430s


eval_pred:   0%|                                                                                | 0/25 [00:00<?, ?it/s]

Epoch 15 elapsed: 257.163s


eval_rec:   0%|                                                                       | 1/1000 [00:00<02:00,  8.32it/s]

{'roc_auc': 0.6810931621735752, 'precision': 0.00575, 'recall': 0.02}


eval_rec: 100%|████████████████████████████████████████████████████████████████████| 1000/1000 [02:02<00:00,  8.16it/s]

recall @90 {'recall': 0.11683333333333334}
Wall time: 1h 19min 1s


In [38]:
ytb_ranking = DeepFM(task="ranking", data_info=data_info,num_neg=3,)

ytb_ranking.fit(train_data, verbose=1, shuffle=True,
                eval_data=test_data,
                metrics=[ "loss", "roc_auc", "precision", "recall"]) # "loss", "roc_auc", "precision", "recall"


print(ytb_ranking.evaluate(test_data,metrics=[ "roc_auc", "precision", "recall"],k=4))
print("recall @90",ytb_ranking.evaluate(test_data,metrics=[ "recall"],k=90))

Training start time: 2021-01-06 22:16:24
total params: 7,161,925 | embedding params: 7,107,140 | network params: 54,785


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:38<00:00, 42.86it/s]


Epoch 1 elapsed: 220.480s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:42<00:00, 42.20it/s]


Epoch 2 elapsed: 223.824s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:44<00:00, 41.73it/s]


Epoch 3 elapsed: 226.317s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:40<00:00, 42.47it/s]


Epoch 4 elapsed: 222.670s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:40<00:00, 42.56it/s]


Epoch 5 elapsed: 221.711s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:35<00:00, 43.62it/s]


Epoch 6 elapsed: 216.292s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:38<00:00, 43.00it/s]


Epoch 7 elapsed: 219.579s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:36<00:00, 43.26it/s]


Epoch 8 elapsed: 218.272s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:36<00:00, 43.35it/s]


Epoch 9 elapsed: 217.858s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:37<00:00, 43.20it/s]


Epoch 10 elapsed: 218.409s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:39<00:00, 42.66it/s]


Epoch 11 elapsed: 221.279s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:38<00:00, 43.01it/s]


Epoch 12 elapsed: 219.440s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:37<00:00, 43.05it/s]


Epoch 13 elapsed: 219.386s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:25<00:00, 45.65it/s]


Epoch 14 elapsed: 206.842s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:30<00:00, 44.67it/s]


Epoch 15 elapsed: 211.382s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:31<00:00, 44.33it/s]


Epoch 16 elapsed: 212.857s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:14<00:00, 48.19it/s]


Epoch 17 elapsed: 195.858s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:14<00:00, 48.25it/s]


Epoch 18 elapsed: 195.596s


train: 100%|███████████████████████████████████████████████████████████████████████| 9382/9382 [03:14<00:00, 48.26it/s]


Epoch 19 elapsed: 195.527s


eval_pred:  12%|████████▋                                                               | 3/25 [00:00<00:00, 24.06it/s]

Epoch 20 elapsed: 201.223s


eval_rec:   1%|▋                                                                     | 10/1000 [00:00<00:10, 97.35it/s]

{'roc_auc': 0.6412648057996093, 'precision': 0.00175, 'recall': 0.0065}


eval_rec: 100%|████████████████████████████████████████████████████████████████████| 1000/1000 [00:10<00:00, 95.52it/s]

recall @90 {'recall': 0.09833333333333334}


### temporal split, min 25:   (results the same on random split ) 
train_data.build_negative_samples(data_info,num_neg=7)
test_data.build_negative_samples(data_info,num_neg=1)

train_data, test_data = split_by_ratio_chrono(df, test_size=0.15)
ytb_ranking = YouTubeRanking(task="ranking", data_info=data_info,
                             embed_size=16, n_epochs=7, lr=2e-4,
                             batch_size=64, use_bn=True,
                             hidden_units="128,64")
============================== YouTubeRanking ==============================
trainset made
testset made
random neg item sampling elapsed: 1.895s
train negs done
random neg item sampling elapsed: 0.080s
test negs done
n_users: 162904, n_items: 3667, data sparsity: 0.0977 %
Training start time: 2021-01-06 00:43:33
train: 100%| 
Epoch 7 elapsed: 1096.930s
Wall time: 2h 15min 44s

Eval (temporal split): 
* {'loss': 0.806, 'roc_auc': 0.982, 'precision': 0.0885, 'recall': 0.3468}
* {recall @ 90 : 'recall': 0.771}


On a random split with same model (might include train on train )  - results are same as with temporal split ; 
* {'loss': 0.806, 'roc_auc': 0.982, 'precision': 0.0885, 'recall': 0.3468}
* recall @90 {'recall': 0.77}

* autoINT (min 25):

{'roc_auc': 0.6810, 'precision': 0.00575, 'recall': 0.02}
recall @90 {'recall': 0.1168}

In [37]:
# reset_state("YouTubeRanking")
# data = pd.read_csv("./sample_movielens_merged.csv", sep=",", header=0)
# data["label"] = 1  # convert to implicit data and do negative sampling afterwards

# # split into train and test data based on time
# train_data, test_data = split_by_ratio_chrono(data, test_size=0.2)

# # specify complete columns information
# sparse_col = ["sex", "occupation", "genre1", "genre2", "genre3"]
# dense_col = ["age"]
# user_col = ["sex", "age", "occupation"]
# item_col = ["genre1", "genre2", "genre3"]

# train_data, data_info = DatasetFeat.build_trainset(
#     train_data, user_col, item_col, sparse_col, dense_col
# )
# test_data = DatasetFeat.build_testset(test_data)
# train_data.build_negative_samples(data_info)  # sample negative items for each record
# test_data.build_negative_samples(data_info)
# print(data_info)  # n_users: 5962, n_items: 3226, data sparsity: 0.4185 %

# ytb_ranking = YouTubeRanking(task="ranking", data_info=data_info, embed_size=16, 
#                              n_epochs=3, lr=1e-4, batch_size=512, use_bn=True, 
#                              hidden_units="128,64,32")
# ytb_ranking.fit(train_data, verbose=2, shuffle=True, eval_data=test_data,
#                 metrics=["loss", "roc_auc", "precision", "recall", "map", "ndcg"])

# # predict preference of user 1 to item 2333
# print("prediction: ", ytb_ranking.predict(user=1, item=2333))  
# # recommend 7 items for user 1
# print("recommendation(id, probability): ", ytb_ranking.recommend_user(user=1, n_rec=7))